### DISCUSSION TITLE
Using Identity(order=True) generates invalid SQL for postgres

### DISCUSSION THREAD

**Discussion by AbdealiLoKo** [sqlalchemy/sqlalchemy#10203] [Usage Questions]
I believe this is a bug - but reporting it here just in case I am wrong ... as I am using `Identity()` for the first time.

I am trying to use `Identity()` for postgres and oracle in my code.
I generally support: MySQL, sqlite, mssql, oracle, and postgres in my application

Reproducible example:
```py
from sqlalchemy import Table, Identity, Column, BigInteger, MetaData
from sqlalchemy.schema import CreateTable
from sqlalchemy.dialects import sqlite, mysql, oracle, mssql, postgresql
meta = MetaData()
table = Table(
    'mytable',
    meta,
    Column('mycol', BigInteger(), Identity(order=True), primary_key=True),
    sqlite_autoincrement=True
)
print('sqlite', CreateTable(table).compile(dialect=sqlite.dialect()))
print('mysql', CreateTable(table).compile(dialect=mysql.dialect()))
print('oracle', CreateTable(table).compile(dialect=oracle.dialect()))
print('mssql', CreateTable(table).compile(dialect=mssql.dialect()))
print('postgresql', CreateTable(table).compile(dialect=postgresql.dialect()))
```

This gives me the output:
```sql
sqlite
CREATE TABLE mytable (
        mycol BIGINT NOT NULL PRIMARY KEY AUTOINCREMENT
)

mysql
CREATE TABLE mytable (
        mycol BIGINT NOT NULL AUTO_INCREMENT,
        PRIMARY KEY (mycol)
)

oracle
CREATE TABLE mytable (
        mycol NUMBER(19) GENERATED BY DEFAULT AS IDENTITY (ORDER),
        PRIMARY KEY (mycol)
)

mssql
CREATE TABLE mytable (
        mycol BIGINT NOT NULL IDENTITY,
        PRIMARY KEY (mycol)
)

postgresql
CREATE TABLE mytable (
        mycol BIGINT GENERATED BY DEFAULT AS IDENTITY (ORDER),
        PRIMARY KEY (mycol)
)
```

When I run the last one i postgres, it gives the following error:
```
SQL Error [42601]: ERROR: syntax error at or near "ORDER"
  Position: 80
```

I couldn't find any documentation that says `ORDER` is a valid concept in postgres. So, I feel like sqlalchemy is generating it wrongly.
`ORDER` is only valid for oracle (For RAC support)

I am using:
 - python: 3.7.16
 - sqlalchemy: 1.4.47

**Comment by CaselIT** [comment]
Hi,

Yes this seems a bug. I think the main problem here is that identity inherited the options of sequence, and sequence had order, even it's oracle specific.

I would propose the following:
- deprecate `order` in both sequence and identity, replacing it with `oracle_order`. this in v2 only
- ignore order on postgresql. this in 1.4

Are there other options in identity that are backend specific?

**Reply by CaselIT** [reply]
as a workaround until it's fixed an release you can probably use a `before_cursor_execute` event to remove the `(ORDER)` from the `GENERATED BY DEFAULT AS IDENTITY` when the dialect is postgresql

**Reply by AbdealiLoKo** [reply]
I think `oracle_order` would be the clearest too

**Comment by zzzeek** [comment]
"order" is specific to Oracle only for *both* sequence and identity, it's definitely not on any other DB like SQL Server etc for either construct, is that accurate ?

why did we call this "order" in that case?

**Reply by CaselIT** [reply]
not sure. it was added in 1.1.12 for some reason

**Reply by CaselIT** [reply]
https://github.com/sqlalchemy/sqlalchemy/commit/faa6609dac2ce6e55e0f690df3ba88c13133ec5c

**Reply by CaselIT** [reply]
it seems cache is similar (seems only oracle and pg)

**Reply by CaselIT** [reply]
> it seems cache is similar (seems only oracle and pg)

pg docs make no mention of cache as being non stanrdard https://www.postgresql.org/docs/current/sql-createsequence.html 

it's also on mssql, so I'm assuming is actually part of the sql standard https://learn.microsoft.com/en-us/sql/t-sql/statements/create-sequence-transact-sql?view=sql-server-ver16

**Reply by zzzeek** [reply]
sure blame the maintainer :)   

I think what happened is I didnt want to build in the dialect-specific kwarg system into sequence/identity

if we are going to have `oracle_order` then we need to use that system, starting like this:

```diff
diff --git a/lib/sqlalchemy/sql/schema.py b/lib/sqlalchemy/sql/schema.py
index 721b9ee637..fd3eb7dbd8 100644
--- a/lib/sqlalchemy/sql/schema.py
+++ b/lib/sqlalchemy/sql/schema.py
@@ -3595,7 +3595,7 @@ class CallableColumnDefault(ColumnDefault):
             )
 
 
-class IdentityOptions:
+class IdentityOptions(DialectKWArgs):
     """Defines options for a named database sequence or an identity column.
 
     .. versionadded:: 1.3.18
@@ -3616,7 +3616,6 @@ class IdentityOptions:
         nomaxvalue: Optional[bool] = None,
         cycle: Optional[bool] = None,
         cache: Optional[int] = None,
-        order: Optional[bool] = None,
     ) -> None:
         """Construct a :class:`.IdentityOptions` object.
 
@@ -3645,7 +3644,7 @@ class IdentityOptions:
         self.nomaxvalue = nomaxvalue
         self.cycle = cycle
         self.cache = cache
-        self.order = order
+        self._validate_dialect_kwargs(dialect_kw)
 
     @property
     def _increment_is_negative(self) -> bool:

```

**Reply by CaselIT** [reply]
created https://github.com/sqlalchemy/sqlalchemy/issues/10207 so best to move it there 

> sure blame the maintainer :)

in this instance we can blame the contributor 😁

**Reply by AbdealiLoKo** [reply]
It may also be worth it to check on the `on_null` kwarg.

When I test it with:
```py
from sqlalchemy import Table, Identity, Column, BigInteger, MetaData
from sqlalchemy.schema import CreateTable
from sqlalchemy.dialects import sqlite, mysql, oracle, mssql, postgresql
meta = MetaData()
table = Table(
    'mytable',
    meta,
    Column('mycol', BigInteger(), Identity(order=True, on_null=True), primary_key=True),
    sqlite_autoincrement=True
)
print('sqlite', CreateTable(table).compile(dialect=sqlite.dialect()))
print('mysql', CreateTable(table).compile(dialect=mysql.dialect()))
print('oracle', CreateTable(table).compile(dialect=oracle.dialect()))
print('mssql', CreateTable(table).compile(dialect=mssql.dialect()))
print('postgresql', CreateTable(table).compile(dialect=postgresql.dialect()))
```

I can see that only oracle's SQL statement has changed - all others remain unchanged.
So, maybe that is also only in oracle

```
oracle
CREATE TABLE mytable (
        mycol NUMBER(38) GENERATED BY DEFAULT ON NULL AS IDENTITY (ORDER),
        PRIMARY KEY (mycol)
)
```

**Reply by CaselIT** [reply]
seems so, nice find.

I guess that was managed correctly, while order wasn't
